tests/test_requirements.py

```diff
@@ -0,0 +1,178 @@
+
+PYSCRIPT_FOLDER = "tests/test_pyscript_folder"
+
+
+@fixture(autouse=True)
+def bypass_package_install_fixture():
+    """Bypass package installation."""
+    with patch("custom_components.pyscript.requirements.async_process_requirements"):
+        yield
+
+
+async def test_install_requirements(hass, caplog):
+    """Test install_requirements function."""
+    with patch(
+        "custom_components.pyscript.requirements.process_all_requirements"
+    ) as process_requirements, patch(
+        "custom_components.pyscript.requirements.async_process_requirements"
+    ) as ha_install_requirements:
+        entry = MockConfigEntry(domain=DOMAIN, data={})
+        entry.add_to_hass(hass)
+
+        # Check that packages get installed correctly
+        process_requirements.return_value = {
+            "my-package-name": {
+                ATTR_SOURCES: [
+                    f"{PYSCRIPT_FOLDER}/requirements.txt",
+                    f"{PYSCRIPT_FOLDER}/apps/app1/requirements.txt",
+                ],
+                ATTR_VERSION: "2.0.1",
+                ATTR_INSTALLED_VERSION: None,
+            },
+            "my-package-name-alternate": {
+                ATTR_SOURCES: [f"{PYSCRIPT_FOLDER}/requirements.txt"],
+                ATTR_VERSION: "2.0.1",
+                ATTR_INSTALLED_VERSION: None,
+            },
+        }
+        await install_requirements(hass, entry, PYSCRIPT_FOLDER)
+        await hass.async_block_till_done()
+        assert ha_install_requirements.called
+        assert ha_install_requirements.call_args[0][2] == [
+            "my-package-name==2.0.1",
+            "my-package-name-alternate==2.0.1",
+        ]
+        assert CONF_INSTALLED_PACKAGES in entry.data
+        assert entry.data[CONF_INSTALLED_PACKAGES] == {
+            "my-package-name": "2.0.1",
+            "my-package-name-alternate": "2.0.1",
+        }
+
+        # Check that we stop tracking packages whose version no longer matches what
+        # we have stored (see previous line for what we currently have stored)
+        ha_install_requirements.reset_mock()
+        caplog.clear()
+        process_requirements.return_value = {
+            "my-package-name": {
+                ATTR_SOURCES: [
+                    f"{PYSCRIPT_FOLDER}/requirements.txt",
+                    f"{PYSCRIPT_FOLDER}/apps/app1/requirements.txt",
+                ],
+                ATTR_VERSION: "2.0.1",
+                ATTR_INSTALLED_VERSION: "2.0.1",
+            },
+            "my-package-name-alternate": {
+                ATTR_SOURCES: [f"{PYSCRIPT_FOLDER}/requirements.txt"],
+                ATTR_VERSION: "2.0.1",
+                ATTR_INSTALLED_VERSION: "1.2.1",
+            },
+        }
+        await install_requirements(hass, entry, PYSCRIPT_FOLDER)
+        await hass.async_block_till_done()
+        assert not ha_install_requirements.called
+        assert "WARNING" in caplog.text and "installed outside of pyscript" in caplog.text
```

Comment:
This assert could be combined into a single longer string, including the name of the package, so that it's more specific.

When I write these caplog asserts I first put in something that fails.  Based on the log output, I find the string in the log, confirm it's what I expect, and then paste that into the caplog assert statement.